### DISCUSSION TITLE
Proposal: Generic Named Type Arguments

### DISCUSSION THREAD

**Discussion by pedro-gilmora** [dotnet/csharplang#8272] [Language Ideas] [answered]
# Proposal: Generic Named Type Arguments

## Summary
Introduce named type arguments in C# to specify type arguments by name at invocation time. This feature aims to simplify generic method invocations, improving code readability and maintainability.

## Motivation
Currently, when invoking generic methods with multiple type parameters, developers often need to specify all type arguments explicitly. This can lead to verbose and less readable code, especially when type inference is not straightforward. Named type arguments at invocation time would simplify this process by allowing developers to specify only the necessary type arguments by name.

## Proposal
Introduce syntax for named type arguments in C# at invocation time. For example:

### Current state
```csharp
//For this signature 
TIn Process<TIn, TOut>(TIn input);

// Current invocation syntax
string result = Process<int, string>(45);
```

### Possible workaround
A possible workaround would be defining this signature
```csharp
TIn Process<TIn, TOut>(TIn input, TOut output);
```

And call
```csharp
string result = Process(45, default(string));
```

### 🎉 **Proposed syntax with named type arguments will infer `result` as `string`**
```csharp
var result = Process<TOut: string>(45);
```

### Design
1. **Syntax Change**:
   - Allow type arguments to be specified by name using the `:` symbol at invocation time.
   - Named type arguments would infer the types of other parameters based on the specified named type argument.

2. **Compiler Implementation**:
   - Update the C# compiler to recognize and interpret named type arguments at invocation time.
   - Ensure compatibility with existing generic method invocation syntax.

### Benefits
- **Simplified Invocations**: Named type arguments make it clear what each type argument represents without needing to specify all generic type parameters in (maybe long) signatures scenarios where infer type argument is needed.
- **Improved Readability**: Code becomes more readable and concise, especially in scenarios with complex generic type inference.
- **Maintainability**: Reduces the likelihood of errors related to type inference and improves the maintainability of codebases.

### Drawbacks
- **Learning Curve**: Developers would need to learn and adapt to the new syntax, although it simplifies existing patterns and aligns with named parameter conventions.

## Alternatives
- **Status Quo**: Continue using the existing unnamed type argument syntax, which requires specifying all type arguments explicitly.
- **Documentation and Comments**: Encourage better documentation and comments to explain the meaning of type arguments. This helps but does not provide the same level of clarity or simplification as named type arguments.

## Conclusion
Named type arguments in C# at invocation time would significantly improve code readability and maintainability, particularly for generic methods with multiple type parameters. This proposal aims to introduce this feature in a way that is intuitive, consistent with existing language features, and improves developer productivity by simplifying method invocations.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/1349.
